src/cleancall.c

```diff
@@ -35,14 +43,23 @@ static void call_exits(void* data) {
 
     void (*fn)(void*) = (void (*)(void*)) R_ExternalPtrAddrFn(CAR(cb));
     void *data = (void*) EXTPTR_PTR(CDR(cb));
+    int nonlocal = LOGICAL(EXTPTR_TAG(CDR(cb)))[0];
 
     // Check for empty pointer in preallocated callbacks
     if (fn) {
+      if (! nonlocal || ! datawpr->success)
       fn(data);
     }
   }
 }
 
+SEXP r_with_cleanup_context_wrap(void *data) {
```

Comment:
Should be static and not namespaced in `r_`

```suggestion
static SEXP with_cleanup_context_wrap(void *data) {
```